### PR TITLE
add operations test for adv rest

### PR DESCRIPTION
**Pull request by ryanfkeepers**
ads operations level tests for advanced restore
configuration on all three services.
Code is largely boilerplate between each service, but with just enough quirks that full consolidation would require excess jumping through hoops.

---

#### Does this PR need a docs update or release note?

- [x] :no_entry: No

#### Type of change

- [x] :robot: Supportability/Tests

#### Issue(s)

* #3562

#### Test Plan

- [x] :green_heart: E2E

### CODE REVIEW

File: src/internal/operations/test/exchange_test.go [hunk 1]
```diff
@@ -842,3 +845,333 @@ func testExchangeContinuousBackups(suite *ExchangeBackupIntgSuite, toggles contr
...
+	t := suite.T()
+
+	ctx, flush := tester.NewContext(t)
+	defer flush()
+
+	// a backup is required to run restores
+
+	baseSel := selectors.NewExchangeBackup([]string{suite.its.userID})
+	baseSel.Include(
+		// cannot be run, for the same reason as incremental backups
```
Comment (ashmrtn) [inline src/internal/operations/test/exchange_test.go:876]: nit: put the data type that can't be run so the comment is complete without having to determine what the commented out code does
```suggestion
		// Events cannot be run, for the same reason as incremental backups
```

File: src/internal/operations/test/exchange_test.go [hunk 2]
```diff
@@ -842,3 +845,333 @@ func testExchangeContinuousBackups(suite *ExchangeBackupIntgSuite, toggles contr
...
+			t,
+			[][]string{tconfig.M365AcctCredEnvs, storeTD.AWSStorageCredEnvs}),
+	})
+}
+
+func (suite *ExchangeRestoreIntgSuite) SetupSuite() {
+	suite.its = newIntegrationTesterSetup(suite.T())
+}
+
+func (suite *ExchangeRestoreIntgSuite) TestRestore_Run_exchangeWithAdvancedOptions() {
```
Comment (ashmrtn) [inline src/internal/operations/test/exchange_test.go:866]: I think this test is good for checking broad strokes of behavior but it doesn't ensure that corso takes the proper action on a per item basis. For example, assume no prior knowledge about how corso operated under the hood. In that case, an implementation that that looked at folder names and decided to take the collision policy action would also pass this test

To ensure proper behavior for individual items we should add a test that restores a subset of items in a folder to start with and have the different policies restore a (possibly disjoint) set of the remaining items in the folder

For example, backup a folder with items 1, 2, 3, and 4. Start by restoring item 1. Then, for each policy restore the next as-of-yet unrestored item and the first item (so [1, 2], [1, 3]...). Or if you didn't want to deal with the fact that item 1's ID would change for the replace policy, backup a folder with 6 items and restore 3 to start with and do [1, 4], [2, 5], [3, 6] for the different policies

File: src/internal/operations/test/restore_helper_test.go [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+	closer func()
+}
+
+func (rod *restoreOpDependencies) close(
+	t *testing.T,
+	ctx context.Context, //revive:disable-line:context-as-argument
+) {
+	fmt.Printf("\n-----\nREST CLOSING\n-----\n")
+
```
Comment (ashmrtn) [inline src/internal/operations/test/restore_helper_test.go:50]:
```suggestion
```

File: src/internal/operations/test/restore_helper_test.go [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+	}
+
+	if rod.kms != nil {
+		err := rod.kw.Close(ctx)
+		assert.NoErrorf(t, err, "kms close: %+v", clues.ToCore(err))
+	}
+}
+
+// prepNewTestRestoreOp generates all clients required to run a restore operation,
+// returning both a resttore operation created with those clients, as well as
```
Comment (ashmrtn) [inline src/internal/operations/test/restore_helper_test.go:65]:
```suggestion
// returning both a restore operation created with those clients, as well as
```

File: src/internal/operations/test/restore_helper_test.go [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+	err := k.Connect(ctx, repository.Options{})
+	require.NoError(t, err, clues.ToCore(err))
+
+	// kopiaRef comes with a count of 1 and Wrapper bumps it again
+	// we're so safe to close here.
+	rod.closer = func() {
+		err := k.Close(ctx)
+		assert.NoErrorf(t, err, "k close: %+v", clues.ToCore(err))
+	}
```
Comment (ashmrtn) [inline src/internal/operations/test/restore_helper_test.go:97]: no matter what it's safe to close k at the end of the function since it's not a part of the restore dependencies struct
```suggestion
	defer k.Close(ctx)
```

File: src/internal/operations/test/restore_helper_test.go [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+
+	rod.kw, err = kopia.NewWrapper(k)
+	if !assert.NoError(t, err, clues.ToCore(err)) {
+		return operations.RestoreOperation{}, nil
+	}
+
+	rod.kms, err = kopia.NewModelStore(k)
+	if !assert.NoError(t, err, clues.ToCore(err)) {
+		return operations.RestoreOperation{}, nil
```
Comment (ashmrtn) [inline src/internal/operations/test/restore_helper_test.go:106]: this should return `rod` so the wrapper can be closed if it was successfully initialized
```suggestion
		return operations.RestoreOperation{}, rod
```

File: src/internal/operations/test/restore_helper_test.go [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+	t *testing.T,
+	ctx context.Context, //revive:disable-line:context-as-argument
+	ro *operations.RestoreOperation,
+	mb *evmock.Bus,
+	expectItemsWritten int, // can be -1 to skip the check
+	acceptNoData bool,
+) *details.Details {
+	deets, err := ro.Run(ctx)
+	require.NoError(t, err, clues.ToCore(err))
+	require.NotEmpty(t, ro.Results, "the restore had non-zero results")
```
Comment (ashmrtn) [inline src/internal/operations/test/restore_helper_test.go:189]: add a comment about why this check is alright. At first glance it seems like a restore that restores no data would fail this check

File: src/internal/operations/test/restore_helper_test.go [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+
+	if expectItemsWritten >= 0 {
+		require.Equal(t, ro.Results.ItemsWritten, expectItemsWritten, "count of items written")
+	}
+
+	assert.NotZero(t, ro.Results.ItemsRead, "count of items read")
+	assert.NotZero(t, ro.Results.BytesRead, "bytes read")
+	assert.Equal(t, 1, ro.Results.ResourceOwners, "count of resource owners")
+	assert.NoError(t, ro.Errors.Failure(), "incremental non-recoverable error", clues.ToCore(ro.Errors.Failure()))
+	assert.Empty(t, ro.Errors.Recovered(), "incremental recoverable/iteration errors")
```
Comment (ashmrtn) [inline src/internal/operations/test/restore_helper_test.go:213]:
```suggestion
	assert.NoError(t, ro.Errors.Failure(), "non-recoverable error", clues.ToCore(ro.Errors.Failure()))
	assert.Empty(t, ro.Errors.Recovered(), "recoverable/iteration errors")
```

File: src/internal/operations/test/restore_helper_test.go [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+	assert.Equal(t, 1, mb.TimesCalled[events.RestoreStart], "restore-start events")
+	assert.Equal(t, 1, mb.TimesCalled[events.RestoreEnd], "restore-end events")
+
+	return deets
+}
+
+type giicbcker[T any] interface {
```
Comment (ashmrtn) [inline src/internal/operations/test/restore_helper_test.go:220]: this needs a more descriptive name

File: src/internal/operations/test/exchange_test.go [hunk 10]
```diff
@@ -842,3 +845,333 @@ func testExchangeContinuousBackups(suite *ExchangeBackupIntgSuite, toggles contr
...
+		m := filterCollisionKeyResults(
+			t,
+			ctx,
+			userID,
+			cIDs[path.EmailCategory],
+			giicbcker[string](acMail),
+			collKeys)
+		maps.Copy(result, m)
+
+		assert.Len(t, result, 0, "no new items should get added")
```
Comment (ashmrtn) [inline src/internal/operations/test/exchange_test.go:1036]: nit
```suggestion
		assert.Empty(t, result, "no new items should get added")
```

File: src/internal/operations/test/exchange_test.go [hunk 11]
```diff
@@ -842,3 +845,333 @@ func testExchangeContinuousBackups(suite *ExchangeBackupIntgSuite, toggles contr
...
+			t,
+			ctx,
+			userID,
+			cIDs[path.EmailCategory],
+			giicbcker[string](acMail),
+			collKeys)
+		maps.Copy(result, m)
+
+		// TODO: we have the option of modifying copy creations in exchange
+		// so that the results don't collid.  But we haven't made that
```
Comment (ashmrtn) [inline src/internal/operations/test/exchange_test.go:1173]:
```suggestion
		// so that the results don't collide.  But we haven't made that
```

File: src/internal/operations/test/exchange_test.go (hunk 10 above)
Comment (ashmrtn) [reply src/internal/operations/test/exchange_test.go:1036]: also, this test is not sufficient to say we didn't duplicate or replace items since there's no item ID comparison happening

File: src/internal/operations/test/sharepoint_test.go [hunk 12]
```diff
@@ -177,3 +181,260 @@ func (suite *SharePointBackupIntgSuite) TestBackup_Run_sharePointExtensions() {
...
+			t,
+			[][]string{tconfig.M365AcctCredEnvs, storeTD.AWSStorageCredEnvs}),
+	})
+}
+
+func (suite *SharePointRestoreIntgSuite) SetupSuite() {
+	suite.its = newIntegrationTesterSetup(suite.T())
+}
+
+func (suite *SharePointRestoreIntgSuite) TestRestore_Run_sharepointWithAdvancedOptions() {
```
Comment (ashmrtn) [inline src/internal/operations/test/sharepoint_test.go:202]: it seems like much of the later part of this test could be pulled out into a helper function and used to run both OneDrive and SharePoint tests?

File: src/internal/operations/test/exchange_test.go (hunk 2 above)
Comment (ryanfkeepers) [reply src/internal/operations/test/exchange_test.go:866]: I'm a little hesitant to run with that implementation as it ends up being a lot of lift and complexity for a relatively small gain.  The primary difference from restoring distinct subsets with expanding data would be to ensure that the collision case handling can pivot on a _per-item_ basis, as opposed to a _per-restore_ basis.  That's a level of granularity that's easyto guarantee with a unit test or a mocked integration test.  I can go look and see if we're adequately testing for that sort of complexity at a lower level, and add those tests if they're missing.  As for here, in the operations layer, I think it's acceptable to work with broader strokes, provided we have those granular tests covered elsewhere.

File: src/internal/operations/test/restore_helper_test.go (hunk 3 above)
Comment (ryanfkeepers) [reply src/internal/operations/test/restore_helper_test.go:50]: whoops, thanks!

File: src/internal/operations/test/restore_helper_test.go (hunk 4 above)
Comment (ryanfkeepers) [reply src/internal/operations/test/restore_helper_test.go:65]: It's not even just the `t` anymore, I'm getting this with many of my keys. :sob:

File: src/internal/operations/test/restore_helper_test.go (hunk 6 above)
Comment (ryanfkeepers) [reply src/internal/operations/test/restore_helper_test.go:106]: good idea

File: src/internal/operations/test/restore_helper_test.go (hunk 9 above)
Comment (ryanfkeepers) [reply src/internal/operations/test/restore_helper_test.go:220]: Heh, I was hoping the initialism would sufficiently pass for its func, to avoid the long name.  `GetItemsInContainerByCollisionKeyer` here we go!

File: src/internal/operations/test/exchange_test.go [hunk 13]
```diff
@@ -842,3 +845,333 @@ func testExchangeContinuousBackups(suite *ExchangeBackupIntgSuite, toggles contr
...
+	var (
+		restoreCfg = ctrlTD.DefaultRestoreConfig("exchange_adv_restore")
+		sel        = rsel.Selector
+		userID     = sel.ID()
+		cIDs       = map[path.CategoryType]string{
+			path.ContactsCategory: "",
+			path.EmailCategory:    "",
+			path.EventsCategory:   "",
+		}
+		collKeys = map[string]string{}
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/operations/test/exchange_test.go:905]: If there's collisions across data types then we'll fail on checks that assert an equal quantity of details entries and collision keys.  Plus, I don't think we have any real possibility of cross-category collision (currently).  Still, I'll add a note here as a warning, just in case.

File: src/internal/operations/test/exchange_test.go (hunk 10 above)
Comment (ryanfkeepers) [reply src/internal/operations/test/exchange_test.go:1036]: I'll add support for this in a branch, and will link the pr when it's ready.

File: src/internal/operations/test/exchange_test.go [hunk 14]
```diff
@@ -842,3 +845,333 @@ func testExchangeContinuousBackups(suite *ExchangeBackupIntgSuite, toggles contr
...
+			cIDs[path.EmailCategory],
+			giicbcker[string](acMail),
+			collKeys)
+		maps.Copy(result, m)
+
+		assert.Len(t, result, 0, "all items should have been replaced")
+
+		for k, v := range result {
+			assert.NotEqual(t, v, collKeys[k], "replaced items should have new IDs")
+		}
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/operations/test/exchange_test.go:1106]: oops, check.

File: src/internal/operations/test/exchange_test.go [hunk 15]
```diff
@@ -842,3 +845,333 @@ func testExchangeContinuousBackups(suite *ExchangeBackupIntgSuite, toggles contr
...
+			userID,
+			cIDs[path.EmailCategory],
+			giicbcker[string](acMail),
+			collKeys)
+		maps.Copy(result, m)
+
+		// TODO: we have the option of modifying copy creations in exchange
+		// so that the results don't collid.  But we haven't made that
+		// decision yet.
+		assert.Len(t, result, 0, "no items should have been added as copies")
```
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/operations/test/exchange_test.go:1175]: Yup, will also fix this in the next update.

File: src/internal/operations/test/sharepoint_test.go (hunk 12 above)
Comment (ryanfkeepers) [reply src/internal/operations/test/sharepoint_test.go:202]: very possible.  Going to save that for a patch on this PR.

File: src/internal/operations/test/exchange_test.go (hunk 2 above)
Comment (ashmrtn) [reply src/internal/operations/test/exchange_test.go:866]: were you able to find any more granular tests?

File: src/internal/operations/test/exchange_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ashmrtn) [reply src/internal/operations/test/exchange_test.go:1169]: is there a plan for checks like this? The subsequent PRs don't seem to address it and I'm not sure if there's some other plan for checking the items have or haven't been replaced/duplicated as expected

[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/operations/test/exchange_test.go:1175]: subsequent PRs aren't wrapped yet.

File: src/internal/operations/test/exchange_test.go (hunk 2 above)
Comment (ryanfkeepers) [reply src/internal/operations/test/exchange_test.go:866]: Yup, each exchange data category has a per-collision set of tests like [this one](https://github.com/alcionai/corso/blob/main/src/internal/m365/exchange/mail_restore_test.go#L102).  And drive restores have a similar test case [here](https://github.com/alcionai/corso/blob/main/src/internal/m365/onedrive/restore_test.go#L329).  I need to go hook up the new counters to those as assurance.  Otherwise, I think that covers us on the implementation regression in a per-item basis.

File: src/internal/operations/test/exchange_test.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (ryanfkeepers) [reply src/internal/operations/test/exchange_test.go:1175]: @ashmrtn how's this: https://github.com/alcionai/corso/pull/3805?

File: src/internal/operations/test/exchange_test.go (hunk 2 above)
Comment (ashmrtn) [reply src/internal/operations/test/exchange_test.go:866]: after some out of band discussion the plan is to get this PR merged along with ID checks for items. After that we can go back and fill out testing a bit more by adding some test cases where only some of the items being restored have the collision policy applied

**Review by ashmrtn — APPROVED** [review]
thanks for all the work on this PR and for bearing with me through all the different asks I had 😅
